library_generation/generate_repo.py

```diff
@@ -0,0 +1,111 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import click
+import library_generation.utilities as util
+from library_generation.generate_composed_library import generate_composed_library
+from library_generation.model.generation_config import from_yaml
+
+
+@click.group(invoke_without_command=False)
+@click.pass_context
+@click.version_option(message="%(version)s")
+def main(ctx):
+    pass
+
+
+@main.command()
+@click.option(
+    "--generation-config-yaml",
+    required=True,
+    type=str,
+    help="""
+    Path to generation_config.yaml that contains the metadata about
+    library generation
+    """,
+)
+@click.option(
+    "--target-library-api-shortname",
+    required=False,
+    type=str,
+    help="""
+    If specified, only the `library` whose api_shortname equals to
+    target-library-api-shortname will be generated.
+    If not specified, all libraries in the configuration yaml will be generated.
+    """,
+)
+@click.option(
+    "--repository-path",
+    required=False,
+    default=".",
+    type=str,
+    help="""
+    If specified, the generated files will be sent to this location.
+    If not specified, the repository will be generated to the current working
+    directory.
+    """,
+)
+def generate(
+    generation_config_yaml: str,
+    target_library_api_shortname: str,
+    repository_path: str,
+):
+    generate_from_yaml(
+        generation_config_yaml=generation_config_yaml,
+        repository_path=repository_path,
+        target_library_api_shortname=target_library_api_shortname,
+    )
+
+
+def generate_from_yaml(
+    generation_config_yaml: str,
+    repository_path: str,
+    target_library_api_shortname: str = None,
+) -> None:
+    """
+    Parses a config yaml and generates libraries via
+    generate_composed_library.py
+    """
+    config = from_yaml(generation_config_yaml)
+    target_libraries = config.libraries
+    if target_library_api_shortname is not None:
+        target_libraries = [
+            library
+            for library in config.libraries
+            if library.api_shortname == target_library_api_shortname
+        ]
+
+    repo_config = util.prepare_repo(
+        gen_config=config, library_config=target_libraries, repo_path=repository_path
+    )
+
+    for library_path, library in repo_config.libraries.items():
+        print(f"generating library {library.api_shortname}")
+
+        generate_composed_library(
+            config=config,
+            library_path=library_path,
+            library=library,
+            output_folder=repo_config.output_folder,
+            versions_file=repo_config.versions_file,
```

Comment:
From design, we probably want to have a check for the existence of such file. AFAIU all repos have this file so it's a good heuristic to ensure that repository_path is valid

Reply:
The versions.txt is checked in `prepare_repo` method in utilities.py.

Reply:
Ah, I see. Thanks